lib/cache/cache.go

```diff
@@ -570,74 +572,120 @@ func (c *Cache) setReadStatus(ok bool, confirmedKinds map[resourceKind]types.Wat
+func readLegacyListResourcesCache(cache *Cache, resourceType string) (legacyReadGuard[resourceGetter], error) {
 	getResourceReader := func(cacheOK bool) resourceGetter {
 		if cacheOK {
 			return cache.presenceCache
 		}
 		return cache.Config.Presence
 	}
 
-	return readCache(cache, types.WatchKind{Kind: resourceType}, getResourceReader)
+	return legacyReadCache(cache, types.WatchKind{Kind: resourceType}, getResourceReader)
 }
 
-// readCache acquires the cache read lock and uses getReader() to select the appropriate target for read operations
+// acquireReadGuard provides a readGuard that may be used to determine how
+// a cache read should operate. The returned guard *must* be released to prevent deadlocks.
+func acquireReadGuard(cache *Cache, kind types.WatchKind) (readGuard, error) {
+	if cache.closed.Load() {
+		return readGuard{}, trace.Errorf("cache is closed")
+	}
+	cache.rw.RLock()
+
+	if cache.ok {
+		if _, kindOK := cache.confirmedKinds[resourceKind{kind: kind.Kind, subkind: kind.SubKind}]; kindOK {
+			return readGuard{
+				cacheRead: true,
+				release:   cache.rw.RUnlock,
+			}, nil
+		}
+	}
+
+	cache.rw.RUnlock()
+	return readGuard{
+		cacheRead: false,
+		release:   nil,
+	}, nil
+}
+
+// legacyReadCache acquires the cache read lock and uses getReader() to select the appropriate target for read operations
 // on resources of the specified kind. The returned guard *must* be released to prevent deadlocks.
-func readCache[R any](cache *Cache, kind types.WatchKind, getReader func(cacheOK bool) R) (readGuard[R], error) {
+func legacyReadCache[R any](cache *Cache, kind types.WatchKind, getReader func(cacheOK bool) R) (legacyReadGuard[R], error) {
 	if cache.closed.Load() {
-		return readGuard[R]{}, trace.Errorf("cache is closed")
+		return legacyReadGuard[R]{}, trace.Errorf("cache is closed")
 	}
 	cache.rw.RLock()
 
 	if cache.ok {
 		if _, kindOK := cache.confirmedKinds[resourceKind{kind: kind.Kind, subkind: kind.SubKind}]; kindOK {
-			return readGuard[R]{
+			return legacyReadGuard[R]{
 				reader:  getReader(true),
 				release: cache.rw.RUnlock,
 			}, nil
 		}
 	}
 
 	cache.rw.RUnlock()
-	return readGuard[R]{
+	return legacyReadGuard[R]{
 		reader:  getReader(false),
 		release: nil,
 	}, nil
 }
 
-// readGuard holds a reference to a read-only "backend" R. If the referenced backed is the cache, then readGuard
+// legacyReadGuard holds a reference to a read-only "backend" R. If the referenced backed is the cache, then legacyReadGuard
 // also holds the release function for the read lock, and ensures that it is not double-called.
-type readGuard[R any] struct {
+type legacyReadGuard[R any] struct {
 	reader   R
 	release  func()
 	released bool
 }
 
 // Release releases the read lock if it is held.  This method
 // can be called multiple times, but is not thread-safe.
-func (r *readGuard[R]) Release() {
+func (r *legacyReadGuard[R]) Release() {
```

Comment:
nit: I'm not sure when/if it would ever come up, but there's a divergence between `legacyReadGuard.Release` and `readGuard.Release`.  The new one is concurrency-safe, and the old one isn't.  I think if we're going to make the new one concurrency-safe, it would be best to update this one as well.  Since they are so similar, it seems likely folks might accidentally assume they have the same safety guarantees.